postgres/src/transaction-manager.ts

```diff
@@ -1,61 +1,65 @@
-import { PoolClient } from 'pg'
-import { ClientManager } from './client-manager'
-import { rollback, commit, unsafeCommit } from './transactions'
-import { PoolProvider } from './pool-provider'
+import type { PoolClient } from 'pg'
+import uuid from './utils/uuid'
 
-export interface TransactionManagerOptions {
-  prepared?: boolean
+export interface TransactionManager {
+  onConnect(client: PoolClient): Promise<void>
+  onSuccess(clients: PoolClient[]): Promise<void>
+  onError(clients: PoolClient[]): Promise<void>
 }
 
-export class TransactionManager extends ClientManager {
-  #doCommit: typeof commit
-
-  constructor(provider: PoolProvider, options: TransactionManagerOptions = {}) {
-    super(provider)
-    this.#doCommit = (options.prepared ?? true) ? commit : unsafeCommit
+export function getTransactionManager(type = 'default'): TransactionManager {
+  switch (type) {
+    case 'none': return null
+    case 'default': return new DefaultTransactionManager()
+    case 'two-phase': return new TwoPhaseTransactionManager()
+    default: throw TypeError(`Unknown transaction manager type "${type}"`)
   }
+}
 
-  protected async createClient(name: string): Promise<PoolClient> {
-    const client = await super.createClient(name)
+class DefaultTransactionManager implements TransactionManager {
+  async onConnect(client: PoolClient): Promise<void> { // eslint-disable-line class-methods-use-this
     await client.query('BEGIN')
-    return client
   }
 
-  public async transactional<T>(task: () => Promise<T>): Promise<T> {
-    try {
-      const result = await task()
-      await this.commit()
-      return result
-    } catch (err) {
-      await this.rollback()
-      throw err
-    }
+  async onSuccess(clients: PoolClient[]): Promise<void> { // eslint-disable-line class-methods-use-this
+    for (const client of clients) await client.query('COMMIT')
   }
 
-  public async* transactionalYield(yieldValue: any) {
-    try {
-      const result = yield yieldValue
-      await this.commit()
-      return result
-    } catch (err) {
-      await this.rollback()
-      throw err
-    }
+  async onError(clients: PoolClient[]): Promise<void> { // eslint-disable-line class-methods-use-this
+    await Promise.allSettled(clients.map(client => client.query('ROLLBACK')))
   }
+}
 
-  async commit() {
-    try {
-      await this.#doCommit(await this.getClients())
-    } finally {
-      await this.release()
-    }
+class TwoPhaseTransactionManager implements TransactionManager {
+  private transactionsIds = new Map<PoolClient, string>()
+
+  async onConnect(client: PoolClient): Promise<void> { // eslint-disable-line class-methods-use-this
+    await client.query('BEGIN')
   }
 
-  async rollback() {
-    try {
-      await rollback(await this.getClients())
-    } finally {
-      await this.release(true)
+  async onSuccess(clients: PoolClient[]): Promise<void> {
+    // FIXME use Promise.all ?
```

Comment:
No, perhaps a `promise.allSettled`. Because we have to know for each client if the transaction has been prepared or not, the rollback phase depends on it.

Reply:
The idea to do it one transaction at a time was to break at first error and avoid preparing transactions that will be rollback immediatly.